File: src/codegen/subresource/subresource.py
#
# Copyright (c) 2010 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


from codegen.utils.typeutil import TypeUtil
from codegen.doc.documentation import Documentation
from codegen.utils.paramutils import ParamUtils
from ovirtsdk.utils.parsehelper import ParseHelper
from codegen.utils.urlutils import UrlUtils
from codegen.utils.headerutils import HeaderUtils
from codegen.resource.resource import Resource
from codegen.templates.subresourcetemplate import SubResourceTemplate
from codegen.templates.subresourcecollectionactiontemplate import SubResourceCollectionActionTemplate
from codegen.templates.subresourceactiontemplate import SubResourceActionTemplate
from codegen.templates.subresourceupdatetemplate import SubResourceUpdateTemplate
from codegen.templates.subresourcedeletetemplate import SubResourceDeleteTemplate
from codegen.templates.subresourcedeletewithbodytemplate import SubResourceDeleteWithBodyTemplate
from codegen.templates.subresourcedeletewithurlparamstemplate import SubResourceDeleteWithUrlParamsTemplate
from codegen.templates.subresourcedeletewithurlparamsandbodytemplate import SubResourceDeleteWithUrlParamsAndBodyTemplate


#============================================================
#=====================SUB COLLECTION RESOURCE================
#============================================================

subresourcetemplate = SubResourceTemplate()
subresourcecollectionactiontemplate = SubResourceCollectionActionTemplate()
subresourceactiontemplate = SubResourceActionTemplate()
subresourceupdatetemplate = SubResourceUpdateTemplate()
subresourcedeletetemplate = SubResourceDeleteTemplate()
subresourcedeletewithbodytemplate = SubResourceDeleteWithBodyTemplate()
subresourcedeletewithurlparamstemplate = SubResourceDeleteWithUrlParamsTemplate()
subresourcedeletewithurlparamsandbodytemplate = SubResourceDeleteWithUrlParamsAndBodyTemplate()

class SubResource(object):
    @staticmethod
    def resource(sub_res_type, encapsulating_entity, parent, KNOWN_WRAPPER_TYPES={}):

        actual_xml_entity = TypeUtil.getValueByKeyOrNone(
                                     encapsulating_entity.lower(),
                                     KNOWN_WRAPPER_TYPES
                             )

        actual_xml_type = ParseHelper.getXmlType(encapsulating_entity)

        actual_sub_res_type = TypeUtil.getValueByKeyOrNone(
                                     sub_res_type.lower(),
                                     KNOWN_WRAPPER_TYPES
                             )

        sub_collection_resource_template_values = {
           'encapsulating_entity':actual_sub_res_type \
                    if actual_sub_res_type is not None\
                    else sub_res_type,
           'sub_res_type': actual_sub_res_type \
                    if actual_sub_res_type is not None \
                    else actual_xml_entity \
                        if actual_xml_entity is not None \
                        else actual_xml_type.__name__ \
                            if actual_xml_type is not None \
                            else encapsulating_entity,
           'parent':parent.lower(),
           'encapsulated_entity':encapsulating_entity.lower(),
           'fixme':Resource.SUB_COLLECTIONS_FIXME
       }

        return subresourcetemplate\
                .generate(sub_collection_resource_template_values)

    @staticmethod
    def action(url, link, action_name, parent_resource_name_lc,
               body_type, resource_name_lc, method, action_params={},
               collection_action=False):

        headers_method_params_str, headers_map_params_str = \
            HeaderUtils.generate_method_params(link)

        headers_method_params_str = ', ' + headers_method_params_str \
                if headers_method_params_str != '' \
                else headers_method_params_str


        sub_collection_resource_action_template_values = {
              'url':url,
              'action_name':action_name,
              'method': method,
              'parent_resource_name_lc':parent_resource_name_lc.lower(),
              'body_type':body_type,
              'body_type_lc':body_type.lower(),
              'headers_map_params_str' : headers_map_params_str,
              'headers_method_params_str' : headers_method_params_str,
              'resource_name_lc':resource_name_lc.lower(),
              'docs' : Documentation.document(link),
              'add_method_params' : Resource._addMethodParams(
                                        action_params.keys()
                                    ),
              'add_action_parans' : Resource._addActionParams(
                                        action_params
                                    ),
              'url_params' : UrlUtils.generate_url_identifiers_replacments(
                                           link,
                                           offset="                ",
                                           continues=True
                                      )
        }

        if collection_action == True:
            sub_collection_resource_action_template = \
                subresourcecollectionactiontemplate.generate(
                         sub_collection_resource_action_template_values
                )
        else:
            sub_collection_resource_action_template = \
                subresourceactiontemplate.generate(
                        sub_collection_resource_action_template_values
                )

        return sub_collection_resource_action_template

    @staticmethod
    def update(url, link, parent_resource_name_lc, resource_name,
               returned_type, KNOWN_WRAPPER_TYPES):

        actual_xml_entity = TypeUtil.getValueByKeyOrNone(
                                         returned_type.lower(),
                                         KNOWN_WRAPPER_TYPES
                                     )

        headers_method_params_str, headers_map_params_str = \
            HeaderUtils.generate_method_params(link)

        headers_method_params_str = ', ' + headers_method_params_str \
                if headers_method_params_str != '' \
                else headers_method_params_str

        prms_str, method_params, url_params = \
            ParamUtils.getMethodParamsByUrlParamsMeta(link)

        if prms_str != '' or headers_method_params_str != '':
            combined_method_params = \
            ', ' if prms_str != '' and not prms_str.startswith(',') else ''
            combined_method_params += prms_str + headers_method_params_str

        method_params_copy = method_params.copy()

        url_params = ParamUtils.toDictStr(
                                      url_params.keys(),
                                      method_params_copy.keys()
                                )

        url_identifiers_replacments = \
            UrlUtils.generate_url_identifiers_replacments(
                          link,
                          offset="            ",
                          continues=True
                      )

        sub_collection_resource_update_template_values = {
             'url':url,
            'parent_resource_name_lc':parent_resource_name_lc.lower(),
            'resource_name':resource_name,
            'resource_name_lc':resource_name.lower(),
            'url_identifiers_replacments' : url_identifiers_replacments,
            'headers_map_params_str' : headers_map_params_str,
            'combined_method_params' : combined_method_params,
            'url_params_separator' : (", " + url_params) if url_params != '' else '',
            'docs' : Documentation.document(link, {}, method_params),
            'returned_type':actual_xml_entity if actual_xml_entity is not None
                                              else returned_type
        }

        return subresourceupdatetemplate\
                .generate(sub_collection_resource_update_template_values)

    @staticmethod
    def delete(url, link, parent_resource_name_lc, resource_name_lc, body_type):

        prms_str, method_params, url_params = \
            ParamUtils.getMethodParamsByUrlParamsMeta(link)

        headers_method_params_str, headers_map_params_str = \
            HeaderUtils.generate_method_params(link)

        combined_method_params = prms_str + (', ' \
                                     if prms_str != '' and headers_method_params_str != '' \
                                     else '') + headers_method_params_str

        headers_map_params_str_with_no_ct = \
            headers_map_params_str.replace('}', ',"Content-type":None}') \
                                            if headers_map_params_str != '{}' \
                                            else '{"Content-type":None}'

        body_instance = ParamUtils.getBodyInstance(link)
        body_instance_str = '=' + body_instance if body_instance else ''


        sub_collection_resource_delete_template_values = {
              'url':url,
              'body_type_lc':body_type.lower() if body_type is not None
                                               else None,
              'parent_resource_name_lc':parent_resource_name_lc.lower(),
              'headers_map_params_str_with_no_ct' : headers_map_params_str_with_no_ct,
              'combined_method_params' : combined_method_params,
              'resource_name_lc':resource_name_lc.lower(),
              'body_instance_str' : body_instance_str,
              'headers_map_params_str': headers_map_params_str,
              'url_params' : ParamUtils.toDictStr(
                                          url_params.keys(),
                                          method_params.copy().keys()
                                        )
        }

        if prms_str != '' or headers_method_params_str != '':
            sub_collection_resource_delete_template_values['docs'] = \
                Documentation.document(link, {}, method_params)

            sub_collection_resource_delete_template_values['url_identifyiers'] = \
                UrlUtils.generate_url_identifiers_replacments(
                           link,
                           offset="            "
                         )

            sub_collection_resource_delete_template = \
                subresourcedeletewithurlparamstemplate\
                    .generate(sub_collection_resource_delete_template_values)

            body_sub_collection_resource_delete_template = \
                subresourcedeletewithurlparamsandbodytemplate\
                    .generate(sub_collection_resource_delete_template_values)
        else:
            sub_collection_resource_delete_template_values['docs'] = \
                Documentation.document(link)

            sub_collection_resource_delete_template_values['url_identifyiers'] = \
                UrlUtils.generate_url_identifiers_replacments(
                           link,
                           offset="                ",
                           continues=True
                         )
            sub_collection_resource_delete_template_values['url_identifyiers_with_body'] = \
                UrlUtils.generate_url_identifiers_replacments(
                           link,
                           offset="                                                               "
                         )

            sub_collection_resource_delete_template = \
                subresourcedeletetemplate\
                    .generate(sub_collection_resource_delete_template_values)


            body_sub_collection_resource_delete_template = \
                subresourcedeletewithbodytemplate\
                    .generate(sub_collection_resource_delete_template_values)

        if not body_type:
            return sub_collection_resource_delete_template
        else:
            return body_sub_collection_resource_delete_template

File: generator/src/main/java/org/ovirt/engine/sdk/generator/rsdl/SubResource.java
//
// Copyright (c) 2014 Red Hat, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package org.ovirt.engine.sdk.generator.rsdl;

import static org.ovirt.engine.sdk.generator.utils.StringUtils.firstNotNull;

import java.util.LinkedHashMap;
import java.util.Map;

import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceActionTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceCollectionActionTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceDeleteTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceDeleteWithBodyTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceDeleteWithUrlParamsAndBodyTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceDeleteWithUrlParamsTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubResourceUpdateTemplate;
import org.ovirt.engine.sdk.generator.templates.AbstractTemplate;
import org.ovirt.engine.sdk.generator.utils.HeaderUtils;
import org.ovirt.engine.sdk.generator.utils.ParamUtils;
import org.ovirt.engine.sdk.generator.utils.UrlUtils;
import org.ovirt.engine.sdk.generator.xsd.XsdData;
import org.ovirt.engine.sdk.entities.DetailedLink;
import org.ovirt.engine.sdk.entities.HttpMethod;

public class SubResource {
    public static String resource(
        String subResType,
        String encapsulatingEntity,
        String parent,
        Map<String, String> knownWrapperTypes
    )
    {
        String actualXmlEntity = knownWrapperTypes.get(encapsulatingEntity.toLowerCase());

        String actualXmlType = XsdData.getInstance().getXmlType(encapsulatingEntity);

        String actualSubResType = knownWrapperTypes.get(subResType.toLowerCase());

        SubResourceTemplate template = new SubResourceTemplate();
        template.set("encapsulating_entity", firstNotNull(actualSubResType, subResType));
        template.set("sub_res_type", firstNotNull(actualSubResType, actualXmlEntity, actualXmlType, encapsulatingEntity));
        template.set("parent", parent.toLowerCase());
        template.set("encapsulated_entity", encapsulatingEntity.toLowerCase());
        template.set("fixme", Resource.SUB_COLLECTIONS_FIXME);
        return template.evaluate();
    }

    public static String action(
        String url,
        DetailedLink link,
        String actionName,
        String parentResourceNameLc,
        String bodyType,
        String resourceNameLc,
        HttpMethod method,
        Map<String, String> actionParams,
        boolean collectionAction
    )
    {
        String[] result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamsStr = result[0];
        String headersMapParamsStr = result[1];

        headersMethodParamsStr = !headersMethodParamsStr.isEmpty()?
            ", " + headersMethodParamsStr:
            headersMethodParamsStr;

        Map<String, String> values = new LinkedHashMap<>();
        values.put("url", url);
        values.put("action_name", actionName);
        values.put("method", method.toString());
        values.put("parent_resource_name_lc", parentResourceNameLc.toLowerCase());
        values.put("body_type", bodyType);
        values.put("body_type_lc", bodyType.toLowerCase());
        values.put("headers_map_params_str", headersMapParamsStr);
        values.put("headers_method_params_str", headersMethodParamsStr);
        values.put("resource_name_lc", resourceNameLc.toLowerCase());
        values.put("docs", Documentation.document(link));
        values.put("add_method_params", Resource.addMethodParams(actionParams.keySet()));
        values.put("add_action_parans", Resource.addActionParams(actionParams));
        values.put("url_params",
            UrlUtils.generateUrlIdentifiersReplacements(
                link,
                "                ",
                true,
                false
            )
        );

        AbstractTemplate template;
        if (collectionAction) {
            template = new SubResourceCollectionActionTemplate();
        }
        else {
            template = new SubResourceActionTemplate();
        }

        template.set(values);
        return template.evaluate();
    }

    public static String update(
        String url,
        DetailedLink link,
        String parentResourceNameLc,
        String resourceName,
        String returnedType,
        Map<String, String> knownWrapperTypes
    )
    {
        String combinedMethodParams = "";

        String actualXmlEntity = knownWrapperTypes.get(returnedType.toLowerCase());

        Object[] result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamsStr = (String) result[0];
        String headersMapParamsStr = (String) result[1];

        headersMethodParamsStr = !headersMethodParamsStr.isEmpty()?
            ", " + headersMethodParamsStr:
            headersMethodParamsStr;

        result = ParamUtils.getMethodParamsByUrlParamsMeta(link);
        String prmsStr = (String) result[0];
        Map<String, String> methodParams = (Map) result[1];
        Map<String, String> urlParams = (Map) result[2];

        if (!prmsStr.isEmpty() || !headersMethodParamsStr.isEmpty()) {
            combinedMethodParams = !prmsStr.isEmpty() && !prmsStr.startsWith(",")? ", ": "";
            combinedMethodParams += prmsStr + headersMethodParamsStr;
        }

        Map<String, String> methodParamsCopy = new LinkedHashMap<>(methodParams);

        String urlParamsStr = ParamUtils.toDictStr(
            urlParams.keySet(), methodParamsCopy.keySet()
        );

        String urlIdentifiersReplacements = UrlUtils.generateUrlIdentifiersReplacements(
            link,
            "            ",
            true,
            false
        );

        SubResourceUpdateTemplate template = new SubResourceUpdateTemplate();
        template.set("url", url);
        template.set("parent_resource_name_lc", parentResourceNameLc.toLowerCase());
        template.set("resource_name", resourceName);
        template.set("resource_name_lc", resourceName.toLowerCase());
        template.set("url_identifiers_replacments", urlIdentifiersReplacements);
        template.set("headers_map_params_str", headersMapParamsStr);
        template.set("combined_method_params", combinedMethodParams);
        template.set("url_params_separator", !urlParamsStr.isEmpty()? ", " + urlParamsStr: "");
        template.set("docs", Documentation.document(link, new LinkedHashMap<String, String>(), methodParams));
        template.set("returned_type", actualXmlEntity != null? actualXmlEntity: returnedType);

        return template.evaluate();
    }

    public static String delete(
        String url,
        DetailedLink link,
        String parentResourceNameLc,
        String resourceNameLc,
        String bodyType
    )
    {
        Object[] result = ParamUtils.getMethodParamsByUrlParamsMeta(link);
        String prmsStr = (String) result[0];
        LinkedHashMap<String, String> methodParams = (LinkedHashMap) result[1];
        LinkedHashMap<String, String> urlParams = (LinkedHashMap) result[2];

        result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamsStr = (String) result[0];
        String headersMapParamsStr = (String) result[1];

        String combinedMethodParams =
            prmsStr +
            (!prmsStr.isEmpty() && !headersMethodParamsStr.isEmpty()? ", ": "") +
            headersMethodParamsStr;

        String headersMapParamsStrWithNoCt =
            !headersMapParamsStr.equals("{}")?
                headersMapParamsStr.replace("}", ",\"Content-type\":None}"):
                "{\"Content-type\":None}";

        String bodyInstance = ParamUtils.getBodyInstance(link);
        String bodyInstanceStr = !bodyInstance.isEmpty()? "=" + bodyInstance: "";

        Map<String, String> values = new LinkedHashMap<>();
        values.put("url", url);
        values.put("body_type_lc", bodyType != null ? bodyType.toLowerCase() : null);
        values.put("parent_resource_name_lc", parentResourceNameLc.toLowerCase());
        values.put("headers_map_params_str_with_no_ct", headersMapParamsStrWithNoCt);
        values.put("combined_method_params", combinedMethodParams);
        values.put("resource_name_lc", resourceNameLc.toLowerCase());
        values.put("body_instance_str", bodyInstanceStr);
        values.put("headers_map_params_str", headersMapParamsStr);
        values.put("url_params", ParamUtils.toDictStr(urlParams.keySet(), methodParams.keySet()));

        AbstractTemplate templateWithBody;
        AbstractTemplate templateWithoutBody;
        if (!prmsStr.isEmpty() || !headersMethodParamsStr.isEmpty()) {
            values.put("docs", Documentation.document(link, new LinkedHashMap<String, String>(), methodParams));
            values.put("url_identifyiers",
                UrlUtils.generateUrlIdentifiersReplacements(
                    link,
                    "            ",
                    false,
                    false
                )
            );

            templateWithBody = new SubResourceDeleteWithUrlParamsAndBodyTemplate();
            templateWithoutBody = new SubResourceDeleteWithUrlParamsTemplate();
        }
        else {
            values.put("docs", Documentation.document(link));
            values.put("url_identifyiers",
                UrlUtils.generateUrlIdentifiersReplacements(
                    link,
                    "                ",
                    true,
                    false
                )
            );
            values.put("url_identifyiers_with_body",
                UrlUtils.generateUrlIdentifiersReplacements(
                    link,
                    "                                                               ",
                    false,
                    false
                )
            );

            templateWithBody = new SubResourceDeleteWithBodyTemplate();
            templateWithoutBody = new SubResourceDeleteTemplate();
        }

        if (bodyType == null || bodyType.isEmpty()) {
            templateWithoutBody.set(values);
            return templateWithoutBody.evaluate();
        }
        else {
            templateWithBody.set(values);
            return templateWithBody.evaluate();
        }
    }
}
